src/openrct2-ui/windows/ScenarioSelect.cpp

```diff
@@ -594,224 +497,310 @@ class ScenarioSelectWindow final : public Window
-
-    auto lightLineLeftTop2 = ScreenCoordsXY{ strRight, lineY };
-    auto lightLineRightBottom2 = ScreenCoordsXY{ right, lineY };
-    GfxDrawLine(dpi, { lightLineLeftTop2, lightLineRightBottom2 }, lightColour);
-
-    // Draw dark horizontal rule
-    lineY++;
-    auto darkLineLeftTop1 = ScreenCoordsXY{ left, lineY };
-    auto darkLineRightBottom1 = ScreenCoordsXY{ strLeft, lineY };
-    GfxDrawLine(dpi, { darkLineLeftTop1, darkLineRightBottom1 }, darkColour);
-
-    auto darkLineLeftTop2 = ScreenCoordsXY{ strRight, lineY };
-    auto darkLineRightBottom2 = ScreenCoordsXY{ right, lineY };
-    GfxDrawLine(dpi, { darkLineLeftTop2, darkLineRightBottom2 }, darkColour);
-}
+    static bool ScenarioSelectUseSmallFont()
+    {
+        return ThemeGetFlags() & UITHEME_FLAG_USE_ALTERNATIVE_SCENARIO_SELECT_FONT;
+    }
+
+    static int32_t GetScenarioListItemSize()
+    {
+        if (!LocalisationService_UseTrueTypeFont())
+            return TrueFontSize;
+
+        // Scenario title
+        int32_t lineHeight = FontGetLineHeight(FontStyle::Medium);
+
+        // 'Completed by' line
+        lineHeight += FontGetLineHeight(FontStyle::Small);
+
+        return lineHeight;
+    }
+};
 
-static bool IsLockingEnabled(WindowBase* w)
+WindowBase* WindowScenarioselectOpen(scenarioselect_callback callback, bool titleEditor)
 {
-    if (gConfigGeneral.ScenarioSelectMode != SCENARIO_SELECT_MODE_ORIGIN)
-        return false;
-    if (!gConfigGeneral.ScenarioUnlockingEnabled)
-        return false;
-    if (w->selected_tab >= 6)
-        return false;
-    if (_titleEditor)
-        return false;
-
-    return true;
+    return WindowScenarioselectOpen(
+        [callback](std::string_view scenario) { callback(std::string(scenario).c_str()); }, titleEditor);
 }
 
-WindowBase* WindowScenarioselectOpen(std::function<void(std::string_view)> callback, bool titleEditor, bool disableLocking)
+WindowBase* WindowScenarioselectOpen(std::function<void(std::string_view)> callback, bool titleEditor)
 {
-    WindowBase* window;
-
-    _callback = callback;
-    _disableLocking = disableLocking;
+    auto* window = static_cast<ScenarioSelectWindow*>(WindowBringToFrontByClass(WindowClass::ScenarioSelect));
+    if (window != nullptr)
+    {
+        if (window->IsTitleEditor() != titleEditor)
+        {
+            WindowCloseByClass(WindowClass::ScenarioSelect);
+        }
+        else
+        {
+            return window;
+        }
+    }
 
     int32_t screenWidth = ContextGetWidth();
     int32_t screenHeight = ContextGetHeight();
     ScreenCoordsXY screenPos = { (screenWidth - WW) / 2, std::max(TOP_TOOLBAR_HEIGHT + 1, (screenHeight - WH) / 2) };
     window = WindowCreate<ScenarioSelectWindow>(WindowClass::ScenarioSelect, screenPos, WW, WH, 0);
-    window->widgets = window_scenarioselect_widgets;
-    window->highlighted_scenario = nullptr;
-
+    window->Init(callback, titleEditor);
```

Comment:
There are better ways of doing this, either you could implement a constructor, and then move it during `WindowCreate`:
https://github.com/OpenRCT2/OpenRCT2/blob/e1f2de988eb3f4b76bd19c233631d24db730e180/src/openrct2-ui/windows/SavePrompt.cpp#L241-L244
or with the variadic option (ctor arguments coming last):
https://github.com/OpenRCT2/OpenRCT2/blob/056af36cdfefbdb42f78dbbca96a6a20fe4423b6/src/openrct2-ui/windows/LoadSave.cpp#L1053-L1054

Reply:
For all the other windows we have used an Init function for now at least. Perhaps on a second pass